File: DS/SegmentTree.java
package DS;

public class SegmentTree {
    class Segment{
        int start;
        int end;
        int sum;
        Segment left;
        Segment right;
        Segment(int i,int j,int s){
            this.start = i;
            this.end = j;
            this.sum = s;
        }
        Segment(int i,int j,int s,Segment l,Segment r){
            this(i, j, s);
            left = l;
            right = r;
        }
    }
    Segment root;
    Segment build(int i,int j,int[] nums){
        if(i>j) return null;
        if(i==j) return new Segment(i,j,nums[i]);
        int mid = (i+j)/2;
        Segment left = build(i, mid, nums);
        Segment right = build(mid+1, j, nums);
        return new Segment(i, j,left.sum+right.sum,left,right);
    }
    public SegmentTree(int[] nums){
        root = build(0, nums.length-1, nums);
    }
    private int _recQuery(Segment segment,int l,int r){
        if(segment==null || l>r) return 0;
        if(segment.start==l && segment.end==r) return segment.sum;
        int mid = (segment.start+segment.end)/2;
        return _recQuery(segment.left, l, Math.min(r, mid)) + _recQuery(segment.right, Math.max(mid+1, l), r);
    }
    public int querySum(int l,int r){
        return _recQuery(root, l, r);
    }
    private void _recUpdate(Segment segment,int i,int val){
        if(segment.start==i && segment.end==i) {
            segment.sum = val;
            return;
        }
        int mid = (segment.start+segment.end)/2;
        if(i<=mid) _recUpdate(segment.left, i, val);
        else _recUpdate(segment.right, i, val);
        segment.sum = segment.left.sum + segment.right.sum;

    }
    public void update(int ind,int val){
        _recUpdate(root, ind, val);
    }
    public static void main(String[] args) {
        SegmentTree sg = new SegmentTree(new int[]{1,2,3,4,5,6,7,8,9});
        for (int i = 0; i < 9; i++) {
            sg.update(i, -1);
            System.out.println(sg.querySum(0, i));

        }

    }
}

File: DS/segment_tree.py
from typing import List

class RangeQuery :
    def __init__(self,n:int) :
        # approx power of 2
        self.n =  2**(n-1).bit_length()
        self.tree = [0] * (2 * self.n + 1) # 1 based indexing 
        self.max_tree = [0]*(2 * self.n + 1)
        
    def _parent(self,i:int) -> int:
        return i//2;
    def _childs(self,i:int) -> List[int]:
        return [2*i,2*i+1];
    
    def update(self,index:int,diff:int) -> None:
        i = index+self.n # node in the tree
        self.tree[i] +=diff
        self.max_tree[i] += diff
        i = self._parent(i)
        while(i>0):
            left,right = self._childs(i);
            self.tree[i] = (self.tree[left]+self.tree[right])
            self.max_tree[i] = max(self.max_tree[left],self.max_tree[right])
            i = self._parent(i)
        
    def _query_sum(self,ns:int,ne:int,start:int,end:int,node:int) -> int:
        if(ns==start and ne==end): # perfect match
            return self.tree[node] 
        if(ne < start or end<ns): return 0 # doesn't lie in the range
        mid = (ns+ne)//2;
        left,right = self._childs(node)
        return self._query_sum(ns,mid,start,min(mid,end),left) + self._query_sum(mid+1,ne,max(start,mid+1),end,right)
    def query_sum(self,left,right):
        return self._query_sum(0,self.n-1,left,right,1);
    
    def _query_max(self,ns:int,ne:int,start:int,end:int,node:int) -> int:
        if(ns==start and ne==end): # perfect match
            return self.max_tree[node] 
        if(ne < start or end<ns): return float('-inf') # doesn't lie in the range
        mid = (ns+ne)//2;
        left,right = self._childs(node)
        return max(self._query_max(ns,mid,start,min(mid,end),left),self._query_max(mid+1,ne,max(start,mid+1),end,right))
    def query_max(self,left,right):
        return self._query_max(0,self.n-1,left,right,1);
    def _lower_max(self, ns: int, ne: int, k: int, node: int):
        if (self.max_tree[node] < k):
            return float('inf')  # not found in this range
        if (ns == ne):
            return ns # found;
        mid = (ns+ne)//2
        l, r = self._childs(node)
        if (self.max_tree[l] >= k):
            return self.lower_max(ns, mid, k, l)
        return self.lower_max(mid+1, ne, k, r)
    
    """
    Find the index of the first element that is greater than or equal to k.
    @param k: the max value to search for
    @return: the index of the first element that is greater than or equal to k.
    
    """
    def lower_max(self, k:int):
        return self._lower_max(0, self.n-1, k, 1)    
def main():
    n = 5                                       
    arr = [-1,-1,-1,-4,-5]
    rq = RangeQuery(100)
    for i in range(n):
        rq.update(i,arr[i])
    print(rq.query_sum(2,4))
    rq.update(0,10)
    print(rq.query_max(2,4))
    
if __name__ == "__main__":
    main()
    